src/main/java/org/opensrp/web/rest/PlanResource.java

```diff
@@ -127,4 +141,45 @@ public void setPlanService(PlanService planService) {
 			return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
 		}
 	}
+
+	/**
+	 * This method provides an API endpoint that searches for plans using a list of provided
+	 * plan identifiers and returns a subset of fields determined by the list of provided fields
+	 * If no plan identifier(s) are provided the method returns all available plans
+	 * If no fields are provided the method returns all the available fields
+	 * @param identifiers list of plan identifiers
+	 * @param fields list of fields to return
+	 * @return plan definitions whose identifiers match the provided params
+	 */
+	@RequestMapping(value = "/findByIdsWithOptionalFields", method = RequestMethod.GET, produces = { MediaType.APPLICATION_JSON_VALUE })
+	public ResponseEntity<String> findByIdentifiersReturnOptionalFields(HttpServletRequest request,
+			@RequestParam(value = IDENTIFIERS) List<String> identifiers,
+			@RequestParam(value = FIELDS, required = false) List<String> fields){
+		try {
+
+			if (fields != null && !fields.isEmpty()) {
+				for (String fieldName : fields) {
+					if (!doesObjectContainField(new PlanDefinition(),fieldName)) {
+						return new ResponseEntity<>("One or more fields are invalid", HttpStatus.BAD_REQUEST);
+					}
+				}
+			}
+			return new ResponseEntity<>(gson.toJson(
+					planService.getPlansByIdsReturnOptionalFields(identifiers, fields)),
+					RestUtils.getJSONUTF8Headers(), HttpStatus.OK);
+		} catch (Exception e) {
+			logger.error(e.getMessage(), e);
+			return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
+		}
+	}
+
+	public boolean doesObjectContainField(Object object, String fieldName) {
+		Class<?> objectClass = object.getClass();
```

Comment:
This method will not able to validate against properties that have been tagged with @SerializedName annotations